### PR TITLE
fix(aws-ecs-pattern): allow ScheduledTaskBase to run on a public subnet

### LINKED ISSUE #6312: aws-ecs-patterns: Allow ScheduledFargateTask and ScheduledEc2Task to run on a public subnet
Allow `ScheduledFargateTask` and `ScheduledEc2Task` to run in a *public* subnet via a configuration option.

### Use Case

I'm currently using the `ScheduledFargateTask` class to run a number of simple tasks. The tasks need **access to** the internet, but there's no need for them to **be accessible from** the internet. I could run the tasks on a private subnet, but this would mean I would need either a NAT Gateway (expensive) or run a NAT Instance on EC2 (maintenance/complexity overhead).

Since the tasks only run for a few minutes every week I'm willing to sacrifice the extra security that a private subnet provides in favour of a simpler/cheaper system where the tasks run on a public subnet.

However, currently `ScheduledFargateTask` will only run a task if its VPC has a private subnet - if there is no private subnet available, an error is reported. I would like to be able to run my tasks on a VPC defined along the following lines:

```typescript
    const vpc = new ec2.Vpc(stack, 'Vpc', {
      maxAzs: 1,
      natGateways: 0,
      subnetConfiguration: [
        { name: 'public', cidrMask: 24, subnetType: SubnetType.PUBLIC }
      ],
    });
    const cluster = new ecs.Cluster(stack, 'EcsCluster', { vpc });
```

### Proposed Solution

Currently [the `addTaskDefinitionToEventTarget ` method on `ScheduledTaskBase` instantiates the underlying `EcsTask`](https://github.com/aws/aws-cdk/blob/a97b48ae4a7dafe050997e00bb57bf64117b551b/packages/%40aws-cdk/aws-events-targets/lib/ecs-task.ts#L113), but there is no way to specify the `subnetSelection` property of that `EcsTask` and so [it defaults to `{ subnetType: ec2.SubnetType.PRIVATE }`](https://github.com/aws/aws-cdk/blob/a97b48ae4a7dafe050997e00bb57bf64117b551b/packages/%40aws-cdk/aws-ecs-patterns/lib/base/scheduled-task-base.ts#L128-L132).

I propose that we add a `subnetSelection` property to the `ScheduledTaskBaseProps` interface and supply that property when instantiating the underlying `EcsTask`. This new property would default to `{ subnetType: ec2.SubnetType.PRIVATE }` thus retaining the existing behaviour.

### Other

* In [this comment](https://github.com/aws/aws-cdk/issues/1305#issuecomment-514805806) @skinny85 mentioned that this is a gap in the `ScheduledFargateTask` construct.

* In [this comment](https://github.com/aws/aws-cdk/issues/1305#issuecomment-461138871) @moofish32 suggested that using only a public subnet is a reasonable option in some circumstances.

* Reading between the lines, it sounds as if others commenting on [this issue](https://github.com/aws/aws-cdk/issues/1305) might welcome this functionality.

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by floehopper**
Allow instances of ScheduledFargateTask and ScheduledEc2Task to run in a *public* subnet via a configuration option.

The default remains that such instances are restricted to run on private subnets, but it is now possible to allow them to run on public subnets if the user is willing to sacrifice the extra security that a private subnet provides in favour of a simpler/cheaper system that does not require a NAT gateway or a NAT instance.

The new unit test schedules a Fargate task to run on a container in a VPC with no private subnet. Before the changes to ScheduledTaskBase in this commit, this test caused the following error:

    Error: There are no 'Private' subnet groups in this VPC. Available types: Public

fixes #6312

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by floehopper** [comment]
@SoManyHs @pkandasamy91 Here's the PR I promised to address #6312. This is my first aws-cdk PR, so I'm sorry if I've done anything wrong. I'm a bit unclear about the PR Linter error which suggests I need to make a change to a README. Presumably in this case this would be the README in the aws-ecs-pattern package...? Although it's not obvious to me what I would change in there...?

File: packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.scheduled-fargate-task.ts [hunk 1]
```diff
@@ -251,4 +251,33 @@ export = {
...
+        image: ecs.ContainerImage.fromRegistry('henk'),
+      },
+      subnetSelection: { subnetType: ec2.SubnetType.PUBLIC },
+      schedule: events.Schedule.expression('rate(1 minute)')
+    });
+
+    // THEN
+    expect(stack).to(haveResourceLike('AWS::ECS::TaskDefinition', {
+      ContainerDefinitions: [{ Image: "henk" }]
+    }));
```
Comment (PettitWesley) [inline packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.scheduled-fargate-task.ts]: @floehopper Would it be possible to add an expect statement that checks for the public IP setting?

The other test cases that I see check for a lot more.

**Comment by PettitWesley** [comment]
> Although it's not obvious to me what I would change in there...?

AFAIK, the change should be made here:
https://github.com/aws/aws-cdk/tree/master/packages/%40aws-cdk/aws-ecs-patterns#scheduled-tasks

File: packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.scheduled-fargate-task.ts (hunk 1 above)
Comment (floehopper) [reply packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.scheduled-fargate-task.ts]: I've amended the commit to address your feedback on the unit test by changing the assertion to check that the stack has a AWS::Events::Rule configured [to use the public subnet and to have a public IP address][1]. I hope that makes sense.

In doing this I noticed that a number of the other existing tests are a bit confusing in this regard. e.g. the "Scheduled Fargate Task - with MemoryReservation defined" test doesn't actually specify the container memory or assert that it has been set.

Also I think a number of the other tests could be improved by the use of the `haveResourceLike` matcher to reduce the signal-to-noise ratio, e.g. I don't see the point in asserting the presence of the detailed `LogConfiguration` in every test.

If you think these would be good improvements, I'd be happy to tackle them as a separate PR.

[1]: https://github.com/aws/aws-cdk/pull/6624/commits/f8fdc4fca88a59591c396adfb39dd78d7658213d#diff-64a7aca333fa14032084fbd40ac0e30bR282-R286

**Comment by floehopper** [comment]
@PettitWesley 

Thanks for the feedback. Sorry for the delay in responding. Please see my responses inline below:

> Blocking:
> 
> * The linter is failing. Please either add this change to the ECS patterns readme or change it from "feat" to "fix", which will remove the requirement for a readme update.

I considered adding a new example to the ECS patterns README, but it seemed weird to add an example for such a specific case when other more common cases are not documented. So I've changed the commit note and PR description from "feat" -> "fix" as you suggested.

However, the linter still seems to be complaining. I wonder whether this is because I force-pushed to the remote branch...?

> * Please rebase

Done and force-pushed.

> * Please see my comment on the unit test

I've amended the commit to address your feedback on the unit test by changing the assertion to check that the stack has a AWS::Events::Rule configured to use the public subnet and to have a public IP address. See my response to your specific comment for details.

> Non-Blocking:
> 
> * I think a unit test is sufficient for this change; however, you could also update the existing integ test to include this parameter.

I did look at updating the existing integration test. However, given that this new optional property is not set by default and, in fact, setting it probably isn't the recommended approach, it didn't seem sensible to muddy the waters by changing the existing test.

I did also consider creating a new integration test, but that felt like overkill for such a small change. So on balance, I'm happy that the unit test gives sufficient coverage.

File: packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.scheduled-fargate-task.ts (hunk 1 above)
Comment (PettitWesley) [reply packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.scheduled-fargate-task.ts]: > Also I think a number of the other tests could be improved 

So to be honest I'm very new to the CDK myself, but I agree... I was also a bit confused when I looked at the existing tests.

>  I'd be happy to tackle them as a separate PR.

Go ahead! 🤠

File: packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.scheduled-fargate-task.ts [hunk 2]
```diff
@@ -251,4 +251,47 @@ export = {
...
+    expect(stack).to(haveResourceLike('AWS::Events::Rule', {
+      Targets: [
+        {
+          EcsParameters: {
+            NetworkConfiguration: {
+              AwsVpcConfiguration: {
+                AssignPublicIp: 'ENABLED',
+                Subnets: [
+                  {
+                    Ref: 'VpcPublicSubnet1Subnet5C2D37C4'
```
Comment (PettitWesley) [inline packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.scheduled-fargate-task.ts]: This test looks better to me now. Where does this value originate from though?

**Comment by PettitWesley** [comment]
@pkandasamy91 Changing from "feat" to "fix" hasn't appeased the linter. What should we do?

File: packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.scheduled-fargate-task.ts (hunk 2 above)
Comment (floehopper) [reply packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.scheduled-fargate-task.ts]: I suspect the "Vpc" prefix comes from the string ID argument supplied on [this line][1] and the word "Public" comes from the `name` property supplied on [this line][2]. I believe the rest is generated by the `Vpc` CDK resource or its sub-resources. The suffix seems to be deterministic, i.e. the same every time the test is run. I believe this behaviour is intentional in the CDK to ensure the same underlying CloudFormation resources are identified each time.

All the other tests which make assertions against the `AWS::Events::Rule` check for a private subnet with reference "VpcPrivateSubnet1Subnet536B997A" which is generated in a similar way, i.e. there is precedent for this. I'm not sure there's a good way to make it clear where the reference value comes from. Do you have any ideas?

As per my previous comment, I'd be happy to try to improve these tests as a separate PR, but I'd be quite keen to get this one merged first so I can start using the new option if that's OK.

[1]: https://github.com/aws/aws-cdk/pull/6624/commits/f8fdc4fca88a59591c396adfb39dd78d7658213d#diff-64a7aca333fa14032084fbd40ac0e30bR258
[2]: https://github.com/aws/aws-cdk/pull/6624/commits/f8fdc4fca88a59591c396adfb39dd78d7658213d#diff-64a7aca333fa14032084fbd40ac0e30bR261

Comment (PettitWesley) [reply packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.scheduled-fargate-task.ts]: > there is precedent for this

I think this is fine then. No need to add a comment or anything.

**Comment by floehopper** [comment]
> Unfortunately, I think we can't merge this until the linter passes. I'm not sure how to fix that; I will ask someone on Monday.

👍

**Comment by floehopper** [comment]
@PettitWesley @pkandasamy91 

> Unfortunately, I think we can't merge this until the linter passes. I'm not sure how to fix that; I will ask someone on Monday.

I've investigated this a bit further and I can confirm that:

1. [The latest PR Linter build][1] ran against [the latest/amended commit][2] and failed.
2. When I run `prlint` locally for this PR, it passes with the following output:

```
$ node tools/prlint/index.js mandatoryChanges 6624
Creating un-authenticated GitHub Client
⌛  Fetching PR number 6624
⌛  Fetching files for PR number 6624
⌛  Validating...
✅  Success
```

So I'm really not sure what's going on!

[1]: https://github.com/aws/aws-cdk/pull/6624/checks?check_run_id=507990244
[2]: https://github.com/aws/aws-cdk/pull/6624/commits/f8fdc4fca88a59591c396adfb39dd78d7658213d

**Comment by floehopper** [comment]
I'm going to try clicking the "Update branch" button which will merge the latest changes from `master` into this branch in the hope the new commit might make the PR Linter do the right thing.

**Comment by floehopper** [comment]
@PettitWesley Updating the branch seems to have done the trick with the PR Linter. Any chance of an updated review and a merge? Thanks.

**Pull request merged**